polaris/evaluate/_metric.py

```diff
@@ -34,105 +137,229 @@ class MetricInfo(BaseModel):
+        handling = self.on_error
+        default = self.default
+        if handling == "default" and default is None:
+            raise ValueError("Default value must be provided when on_error is 'default'")
+        return self
+
+    def score(
+        self,
+        y_true: GroundTruth,
+        y_pred: BenchmarkPredictions | None,
+        y_prob: BenchmarkPredictions | None,
+        info: MetricInfo,
+    ) -> float:
+        """Compute the metric.
+
+        Args:
+            y_true: The true target values.
+            y_pred: The predicted target values, if any.
+            y_prob: The predicted target probabilities, if any.
+            info: The metadata for the metric.
+        """
+
+        pred = prepare_predictions(y_pred, y_prob, info.y_type)
+
+        # NOTE (cwognum): We rely on pandas to due the grouping.
+        # This does imply a memory bottleneck.
+        df = y_true.as_dataframe()
+        df[info.y_type] = pred.flatten()
+
+        # Compute the metric for each group
+        scores = []
+
+        for _, group in df.groupby(self.group_by):
+            y_true_group = group[y_true.target_cols[0]].values
+            y_pred_group = group[info.y_type].values
+
+            kwargs = prepare_metric_kwargs(
+                y_true_group,
+                y_pred_group,
+                info.y_type,
+            )
+
+            # A group can be arbitrarily small (e.g. only 1 item), which also has implications
+            # for its target distribution (e.g. only the positive class). These conditions can
+            # lead to errors in the metric computation, which we handle here.
+            try:
+                score = info.fn(**kwargs, **info.kwargs)
+            except Exception:
+                if self.on_error == "ignore":
+                    continue
+                elif self.on_error == "raise":
+                    raise
+                score = self.default
+            scores.append(score)
+
+        # Aggregate the per-group score in a single score.
+        if self.aggregation == "mean":
+            score = np.mean(scores)
+        else:
+            score = np.median(scores)
+        return score
+
+
+class Metric(BaseModel):
+    """
+    A Metric in Polaris.
+
+    Attributes:
+        label: The actual callable that is at the core of the metric implementation.
+        config: For more complex metrics, this object should hold all parameters for the metric.
+    """
+
+    label: MetricLabel
+    config: GroupedMetricConfig | None = None
 
     @property
-    def is_multitask(self) -> bool:
-        """Whether the metric expects a single set of predictions or a dict of predictions."""
-        return self.value.is_multitask
+    def info(self) -> MetricInfo:
+        """The metadata for the metric."""
+        metric_info = DEFAULT_METRICS.get(self.label)
+        if metric_info is None:
+            raise ValueError(f"Metric {self.label} not found in the available metrics.")
```

Comment:
**Comment**
This kind of validation should happen when the object is created, to prevent an invalid state from existing in the first place. Hence my initial suggestion to do this in a validator.